Source/Core/DolphinQt2/FIFOPlayerWindow.cpp

```diff
@@ -0,0 +1,359 @@
+// Copyright 2017 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include "DolphinQt2/FIFOPlayerWindow.h"
+
+#include <QCheckBox>
+#include <QDialogButtonBox>
+#include <QFileDialog>
+#include <QGroupBox>
+#include <QHBoxLayout>
+#include <QLabel>
+#include <QMessageBox>
+#include <QPushButton>
+#include <QSpinBox>
+#include <QVBoxLayout>
+
+#include "Core/Core.h"
+#include "Core/FifoPlayer/FifoDataFile.h"
+#include "Core/FifoPlayer/FifoPlaybackAnalyzer.h"
+#include "Core/FifoPlayer/FifoPlayer.h"
+#include "Core/FifoPlayer/FifoRecorder.h"
+
+#include "DolphinQt2/QtUtils/QueueOnObject.h"
+#include "DolphinQt2/Settings.h"
+
+FIFOPlayerWindow::FIFOPlayerWindow(QWidget* parent) : QDialog(parent)
+{
+  setWindowTitle(tr("FIFO Player"));
+
+  CreateWidgets();
+  ConnectWidgets();
+
+  UpdateInfo();
+  UpdateControls(Core::IsRunning());
+
+  connect(this, &FIFOPlayerWindow::EmulationStarted, this, &FIFOPlayerWindow::OnEmulationStarted);
+  connect(this, &FIFOPlayerWindow::EmulationStopped, this, &FIFOPlayerWindow::OnEmulationStopped);
+
+  FifoPlayer::GetInstance().SetFileLoadedCallback(
+      [this] { QueueOnObject(this, [this] { OnFIFOLoaded(); }); });
+  FifoPlayer::GetInstance().SetFrameWrittenCallback(
+      [this] { QueueOnObject(this, [this] { UpdateInfo(); }); });
+}
+
+FIFOPlayerWindow::~FIFOPlayerWindow()
+{
+  FifoPlayer::GetInstance().SetFileLoadedCallback(nullptr);
+  FifoPlayer::GetInstance().SetFrameWrittenCallback(nullptr);
+}
+
+void FIFOPlayerWindow::CreateWidgets()
+{
+  auto* layout = new QVBoxLayout;
+
+  // Info
+  auto* info_group = new QGroupBox(tr("File Info"));
+  auto* info_layout = new QHBoxLayout;
+
+  m_info_label = new QLabel;
+  info_layout->addWidget(m_info_label);
+  info_group->setLayout(info_layout);
+
+  m_info_label->setFixedHeight(QFontMetrics(font()).lineSpacing() * 3);
+
+  // Object Range
+  auto* object_range_group = new QGroupBox(tr("Object Range"));
+  auto* object_range_layout = new QHBoxLayout;
+
+  m_object_range_from = new QSpinBox;
+  m_object_range_from_label = new QLabel(tr("From:"));
+  m_object_range_to = new QSpinBox;
+  m_object_range_to_label = new QLabel(tr("To:"));
+
+  m_object_range_from->setMinimum(0);
```

Comment:
Do these object range minimum and maximums need to be set here? It looks like the controls are disabled until a FIFO log is loaded, in which case they're set accordingly.